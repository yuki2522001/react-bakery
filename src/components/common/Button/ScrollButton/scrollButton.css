.btn__backToTop {
  position: fixed;
  bottom: 100px;
  right: 20px;
  font-size: 25px;
  background: var(--text-primary);
  color: var(--text-link);
  cursor: pointer;
  border-radius: 50px;
  border: none;
  box-shadow: 0 5px 10px var(--border-btn);
}

.btn__scroll {
  font-size: 25px;
  background: var(--text-primary);
  color: var(--text-link);
  cursor: pointer;
  border-radius: 50px;
  border: none;
  box-shadow: 0 5px 10px var(--border-btn);
}