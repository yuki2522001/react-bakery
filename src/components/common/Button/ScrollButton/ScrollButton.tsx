import React, { memo, useEffect, useState } from "react";
import "./scrollButton.css";

interface ButtonScroll {
  className?: string;
  children: JSX.Element;
  onClick?: () => void;
}

const ScrollButton: React.FC<ButtonScroll> = memo(({ children, className }) => {
  // create state to handle back to top
  const [btnOnTop, setBtnOnTop] = useState<boolean>(true);
  useEffect(() => {
    window.addEventListener("scroll", () => {
      if (window.pageYOffset > 500) {
        setBtnOnTop(btnOnTop);
      } else {
        setBtnOnTop(!btnOnTop);
      }
    });
  }, []);

  // This function will scroll the window to the top
  const scrollToTop = () => {
    window.scrollTo({
      top: 0,
      behavior: "smooth", // for smoothly scrolling
    });
  };

  return (
    <>
      {btnOnTop && (
        <button
          data-testid="back-to-top"
          className={className}
          onClick={scrollToTop}
        >
          {children}
        </button>
      )}
      <div></div>
    </>
  );
});

export default ScrollButton;
