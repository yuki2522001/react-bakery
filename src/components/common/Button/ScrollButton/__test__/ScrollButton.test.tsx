import { fireEvent, render } from "@testing-library/react";
import ScrollButton from "../ScrollButton";

describe("ScrollButton component", () => {
  const spyScrollTo = jest.fn();
  beforeEach(() => {
    Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });
    spyScrollTo.mockClear();
  });

  test("should render button when window.pageYOffset > 500", () => {
    const { getByTestId } = render(
      <ScrollButton
        children={<i className="fa fa-arrow-alt-circle-up"></i>}
        onClick={spyScrollTo}
      />
    );
    const clickBtn = getByTestId("back-to-top");
    fireEvent.scroll(window, { target: { pageYOffset: 700 } });
    fireEvent.click(clickBtn);
    expect(spyScrollTo).toHaveBeenCalled();
  });

  test("should not render button when window.pageYOffset < 500", () => {
    const { getByTestId } = render(
      <ScrollButton
        children={<i className="fa fa-arrow-alt-circle-up"></i>}
        onClick={spyScrollTo}
      />
    );
    const clickBtn = getByTestId("back-to-top");
    fireEvent.scroll(window, { target: { pageYOffset: 300 } });
    fireEvent.click(clickBtn);
    expect(spyScrollTo).not.toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ScrollButton
        children={<i className="fa fa-arrow-alt-circle-up"></i>}
        onClick={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
