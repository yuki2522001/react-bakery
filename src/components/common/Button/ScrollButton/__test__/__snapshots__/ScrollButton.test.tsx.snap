// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollButton component matches snapshot 1`] = `
<DocumentFragment>
  <button
    data-testid="back-to-top"
  >
    <i
      class="fa fa-arrow-alt-circle-up"
    />
  </button>
  <div />
</DocumentFragment>
`;
